File: leetcode/001-Two-Sum/TwoSum_002.py
#@author: cchen
class Solution:
    # @param {integer[]} nums
    # @param {integer} target
    # @return {integer[]}
    def twoSum(self, nums, target):
        idx = sorted(range(len(nums)), key = lambda x:nums[x])
        nums.sort()
        i, j = 0, len(nums)-1
        
        while i < j:
            if nums[i] + nums[j] > target:
                j -= 1
            elif nums[i] + nums[j] < target:
                i += 1
            else:
                if idx[i] > idx[j]:
                    return [idx[j] + 1, idx[i] + 1]
                else:
                    return [idx[i] + 1, idx[j] + 1]

File: leetcode/001-Two-Sum/java/TwoSum_002.java
import java.util.Arrays;
import java.util.Comparator;

class ArrayIndexComparator implements Comparator<Integer> {
    private final int[] array;

    public ArrayIndexComparator(int[] nums) {
        array = nums;
    }

    public int compare(Integer index1, Integer index2) {
        return array[index1] - array[index2];
    }
}


public class Solution {
    public int[] twoSum(int[] nums, int target) {
        int[] res = new int[2];
        Integer[] idx = new Integer[nums.length];
        for (int k = 0; k < idx.length; k++) {
            idx[k] = k;
        }

        Arrays.sort(idx, new ArrayIndexComparator(nums));
        Arrays.sort(nums);

        int i = 0, j = nums.length - 1;
        while (i < j) {
            if (nums[i] + nums[j] < target) {
                i++;
            } else if (nums[i] + nums[j] > target) {
                j--;
            } else {
                if (idx[i] > idx[j]) {
                    res[0] = idx[j] + 1;
                    res[1] = idx[i] + 1;
                } else {
                    res[0] = idx[i] + 1;
                    res[1] = idx[j] + 1;
                }
                break;
            }
        }

        return res;
    }
}
